### ISSUE TITLE
When including GMLodash; project fails to compile in YYC mode

### ISSUE THREAD

**Issue by MikeCaputo** [DatZach/GMLodash#1]
Hello! I've discovered what might be an issue with GMLodash. I haven't been able to find a GM-specific bug for this issue (though I'll continue to find if one exists one).

To create a reproducible setup, I created a new blank project and added the following test object to the starting room:

Create event:
```
/// @description Set up struct for listening

users = [
    { user: "foo", age: 25, active: true },
    { user: "bar", age: 29, active: false },
    { user: "baz", age: 29, active: true }
];
```

Step event:
```
/// @description Listen for keyboard input and use GMLodash

if(keyboard_check_pressed(ord("1"))) {
	var returned = _.find(users, { active: false });
	show_debug_message(string(returned));
}

if(keyboard_check_pressed(ord("2"))) {
	var returned = _.find(users, { active: true });
	show_debug_message(string(returned));
}
```

All this does is listen to the keyboard inputs 1 and 2, and it would console out matches within the `users` array. This works find when I run the game in `VM` mode. But when I switch to `YYC` mode and try run it, it fails to build with the following error:

```
Error : gml_Object_testFindFunction_Step_0(0) : Fatal Error while compiling gml_Object_testFindFunction_Step_0 - bailing details below
Final Compile...System.InvalidOperationException: Stack empty.
   at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   at System.Collections.Generic.Stack`1.Pop()
   at GMAssetCompiler.GML2CPP.(GMLToken , Boolean , StringBuilder , CPPVarRef )
   at GMAssetCompiler.GML2CPP.(GMLToken )
   at GMAssetCompiler.GML2CPP.(GMLToken )
   at GMAssetCompiler.GML2CPP.(GMLToken )
   at GMAssetCompiler.GML2CPP.(GMLToken )
   at GMAssetCompiler.GML2CPP.(GMLToken )
   at GMAssetCompiler.GML2CPP.(GMLToken )
   at GMAssetCompiler.GML2CPP.(GMLToken )
   at GMAssetCompiler.GML2CPP.(GMLToken )
   at GMAssetCompiler.GML2CPP.(GMLToken )
   at GMAssetCompiler.GML2CPP.(GMLToken )
   at GMAssetCompiler.GML2CPP.(GMAssets ,  )
finished.
```

And some additional information: when I comment-out the use use `_.find()` and have nothing in the project using GMLodash - but with GMLodash still included as a separate `script` file - I get the following different set of errors:

```
Build for architecture - x86

Z:/text-gm-lo_2443EE70_611E4F00/text-gm-lodash-yyc/Default/Scripts/llvm-win/gml_Object_testFindFunction_Step_0.gml.cpp:43:66: error: use of undeclared identifier 'kVARID_self_returned'
YYRValue* sself_returned = &((CInstanceBase*)pSelf)->GetYYVarRef(kVARID_self_returned); /* set ContextID to sself_returned */
                                                                 ^
1 error generated.
Z:/text-gm-lo_2443EE70_611E4F00/text-gm-lodash-yyc/Default/Scripts/llvm-win/gml_GlobalScript_gmLodash.gml.cpp:4678:10: error: redefinition of 'local_args' with a different type: 'YYRValue' vs 'YYLocalArgs'
YYRValue local_args;
         ^
Z:/text-gm-lo_2443EE70_611E4F00/text-gm-lodash-yyc/Default/Scripts/llvm-win/gml_GlobalScript_gmLodash.gml.cpp:4666:13: note: previous definition is here
YYLocalArgs local_args( _count, 2, _args );
            ^
Z:/text-gm-lo_2443EE70_611E4F00/text-gm-lodash-yyc/Default/Scripts/llvm-win/gml_GlobalScript_gmLodash.gml.cpp:4762:11: error: no viable overloaded '='
local_args=YYGML_ds_stack_pop((int)(int64)((int)(int64)/* local */local_stack.asReal()));
~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
X://yyc/include\YYGML.h:2328:8: note: candidate function (the implicit copy assignment operator) not viable: no known conversion from 'YYRValue' to 'const YYLocalArgs' for 1st argument
struct YYLocalArgs
       ^
Z:/text-gm-lo_2443EE70_611E4F00/text-gm-lodash-yyc/Default/Scripts/llvm-win/gml_GlobalScript_gmLodash.gml.cpp:4765:153: error: type 'YYLocalArgs' does not provide a subscript operator
local_value=/* array index changed from System.Collections.Generic.List`1[System.String] to System.Collections.Generic.List`1[System.String]*/local_args[(int)(0)];
                                                                                                                                              ~~~~~~~~~~^~~~~~~~~
Z:/text-gm-lo_2443EE70_611E4F00/text-gm-lodash-yyc/Default/Scripts/llvm-win/gml_GlobalScript_gmLodash.gml.cpp:4769:37: error: type 'YYLocalArgs' does not provide a subscript operator
local_limit=/* volatile */local_args[(int)(1)];
                          ~~~~~~~~~~^~~~~~~~~
4 errors generated.
```

I'm running GM v2.3.0.529, Runtime v2.3.0.401; I'll continue to look for other causes of this, and I'll update this issue if/when I find any other info.

**Comment by MikeCaputo** [comment]
Update: I'm hot on the trail of this issue. I'll make another update with more information, and hopefully a solution, tomorrow.

**Comment by DatZach** [comment]
Hello,

Can you send me a test project that reproduces this? I’ll make sure it gets seems by YYG so it can be corrected.

Any code that compiles and runs in VM should also do the same in YYC.

Not sure what exactly is happening but it’s definitely on YYG to handle.

Cheers,
Zach Reedy

iPhoneから送信

11/14/20 午後10:31、Mike Caputo <notifications@github.com>のメール:

﻿

Update: I'm hot on the trail of this issue. I'll make another update tomorrow.

**Comment by MikeCaputo** [comment]
Hi Zach,

Attached is my small test project for validating this issue. I've uncovered _one_ part of the issue, but then there is a second issue that I haven't been able to crack. Here is the summary:

1. If you open and run this project, it has GMLodash, and a simple object. Pressing `1` or `2` will run the `_.find()` function and will give different ouputs. **This will run fine in VM mode.**
2. If you switch to YYC mode and try to run it, it will fail.
3. If you comment-out the two instances of `_.find()` being used (and the related logging), and you try to build the project, _it will still fail_ even though GMLodash isn't being called.
4. I was able to get one step further: I stepped through the lodash scripts, looking for the exact point at which GM was failing to build it in YYC mode. I discovered that within the `join` function, GM was failing to interpret the variable name `args` (`var args = ds_stack_pop(stack);`). If you rename the three instances of `args` to anything else (such as `arguments`), and you build the project, then it will build properly. My best guess is that under the hood there is some kind of naming conflict in the compiler.
5. Back in the test object, however, if you then un-comment out the two instances of `_.find()`, then GM throws a different set of errors. I haven't been able to get past these - it seems to be referencing a ` Stack empty`, but I looked through the GMLodash code and it seems to be safely using the stacks. I could be missing something, or this could be a bug deep in the guts of GM which I'm not in a position to diagnose correctly.

Let me know if I can give any other information. Thanks!

[test-gm-lodash-yyc.zip](https://github.com/DatZach/GMLodash/files/5542721/test-gm-lodash-yyc.zip)

**Comment by MikeCaputo** [comment]
And let me know if you'd like me to submit a PR for changing the `args` variable name.

**Comment by DatZach** [comment]
Yes please! I will approve if you submit one.
________________________________

**Comment by MikeCaputo** [comment]
It looks like this _was_ a GM issue, and thankfully it's been fixed in the latest release (`2.3.1.536`): https://gms.yoyogames.com/ReleaseNotes.html

The specific issue was: `Building Projects: [YYC] Referencing a global function inside another constructor/function fails the build`

After upgrading to this latest version, GMLodash compiles and works correctly in YYC mode. Closing this issue!
